sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -568,6 +568,29 @@ public Duration getAllowedTimestampSkew() {
   public @interface ProcessElement {}
 
   /**
+   * <b><i>Experimental - no backwards compatibility guarantees. The exact name or usage of this
+   * feature may change.</i></b>
+   *
+   * <p>Annotation that may be added to a {@link ProcessElement} or {@link OnTimer} method to
+   * indicate that the observable contents of the input {@link PCollection} or mutable state must be
+   * stable upon retries.
+   *
+   * <p>This is important for exactly-once semantics when writing to a storage medium outside of
+   * your pipeline. The general pattern is to establish an idempotent protocol for writing the data,
+   * and then require stable input. Combined, these allow the write to be freely retried until
+   * success.
+   *
+   * <p>An example of an unstable input would be anything computed using nondeterministic logic. In
+   * Beam, any user-defined function is permitted to be nondeterministic, and any {@link
+   * PCollection} is permitted to be recomputed in any manner.
+   */
+  @Documented
+  @Experimental
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target(ElementType.METHOD)
+  public @interface RequiresStableInput {}
```

Comment:
The naming '`RequiresStableInput`' makes it sound like the consumer must ensure stable input. But my understanding is that this requirement is pushed onto runners to ensure stability via checkpointing, correct?

If there's some mechanism for transform authors to `@CreateStableInput`/`@PreserveStableInput`, it should be noted here. Otherwise consider adding a comment that this requirement is targeted towards runners.

Reply:
Added some verbiage that indicates the requirement is on the runner. WDYT?

Reply:
Would it make sense to make this just a parameter of `@ProcessElement`?

`@ProcessElement(onStableInput = true)`

Either way, the important point is, `RequireStableInput` is an annotation a user puts on their processing logic to indicate that it must observe the same input elements.

Reply:
I think you could go either way, but I think this one is a bit more flexible - no need to add the flag to various places you might add it.